### PR TITLE
API: Make Series.searchsorted return a scalar, when supplied a scalar

### PR DESCRIPTION
**Pull request by topper-123**
- [x] xref #23466 & #22034
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

``Series.searchsorted`` returns the wrong shape for scalar input. Numpy arrays and all other array types return a scalar if the input is a scalar, but ``Series`` does not.

For example:

```python
>>> import numpy as np
>>> np.array([1, 2, 3]).searchsorted(1)
0
>>> np.array([1, 2, 3]).searchsorted([1])
array([0])
>>> import pandas as pd
>>> d = pd.date_range('2018', periods=4)
>>> d.searchsorted(d[0])
0
>>> d.searchsorted(d[:1])
array([0])

>>> n = 100_000
>>> s = pd.Series(([0]*n + [1]*n + + [2]*n))
>>> s.searchsorted(1)
array([100000], dtype=int32)  # master
100000  # this PR. Scalar input should lead to scalar output
>>> s.searchsorted([1])
array([100000], dtype=int32)  # master and this PR
```

### CODE REVIEW

File: pandas/util/testing.py [hunk 1]
```diff
@@ -27,7 +27,8 @@
     is_bool, is_categorical_dtype, is_datetime64_dtype, is_datetime64tz_dtype,
     is_datetimelike_v_numeric, is_datetimelike_v_object,
     is_extension_array_dtype, is_interval_dtype, is_list_like, is_number,
-    is_period_dtype, is_sequence, is_timedelta64_dtype, needs_i8_conversion)
+    is_period_dtype, is_scalar, is_sequence, is_timedelta64_dtype,
+    needs_i8_conversion)  # noqa
```
Comment (jreback) [inline pandas/util/testing.py]: why did you change this?

File: doc/source/whatsnew/v0.24.0.rst [hunk 2]
```diff
@@ -1008,6 +1008,7 @@ Other API Changes
 - Slicing a single row of a DataFrame with multiple ExtensionArrays of the same type now preserves the dtype, rather than coercing to object (:issue:`22784`)
 - :class:`DateOffset` attribute `_cacheable` and method `_should_cache` have been removed (:issue:`23118`)
 - Comparing :class:`Timedelta` to be less or greater than unknown types now raises a ``TypeError`` instead of returning ``False`` (:issue:`20829`)
+- :meth:`Series.searchsorted`, when supplied a scalar value to search for, now returns a scalar instead of an array (:issue:`xxxxx`).
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.rst]: use this PR number

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/base.py [hunk 3]
```diff
@@ -1207,8 +1207,14 @@ def factorize(self, sort=False, na_sentinel=-1):
...
         Returns
         -------
-        indices : array of ints
-            Array of insertion points with the same shape as `value`.
+        int or array of ints
+            A scalar or array of insertion points with the
+            same shape as `value`.
+
+            .. versionchanged :: 0.24.0
+                Ìf `value`is a scalar, an int is now always returned.
```
Comment (jschendel) [inline pandas/core/base.py]: - "Ìf" -- > "If"
- missing a space between "\`value\`" and "is"

File: pandas/util/testing.py (hunk 1 above)
Comment (topper-123) [reply pandas/util/testing.py]: ´´is_scalar`` wasn't in testing, I need it in my tests and think it's helpful to have in this common testing module.

I need to to for scalar because ``assert np.array([2]) == 2`` passes without an AssertionError...

Comment (topper-123) [reply pandas/util/testing.py]: Though I do get an flake8 error for unused import. Trying to resolve that...

File: pandas/tests/arrays/categorical/test_analytics.py [hunk 4]
```diff
@@ -86,9 +86,11 @@ def test_searchsorted(self):
         # Searching for single item argument, side='left' (default)
         res_cat = c1.searchsorted('apple')
         assert res_cat == 2
+        assert tm.is_scalar(res_cat)
```
Comment (topper-123) [inline pandas/tests/arrays/categorical/test_analytics.py]: ``tm.is_scalar`` guards against res_cat being an array, as ``assert np.array(...) == 2`` would always pass the assertion test, as long as the array is non-empty.

Another way would be to to do ``assert (res_cat == 2) is True``.

**Comment by topper-123** [comment]
TheTravis failure is unrelated (hypothesis related)

```
>   @given(index=indices(max_length=5), num_columns=integers(0, 5))
>   def test_frequency_is_original(self, index, num_columns):
pandas/tests/frame/test_apply.py:1160: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../../miniconda3/envs/pandas/lib/python2.7/site-packages/hypothesis/core.py:638: in execute
    ) % (test.__name__, text_repr[0],))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
self = <hypothesis.core.StateForActualGivenExecution object at 0x7f2c035ae050>
message = "Hypothesis test_frequency_is_original(self=<pandas.tests.frame.test_apply.TestDataFrameAggregate instance at 0x7f2c3f...', freq=None), num_columns=3) produces unreliable results: Falsified on the first call but did not on a subsequent one"
    def __flaky(self, message):
        if len(self.falsifying_examples) <= 1:
>           raise Flaky(message)
E           Flaky: Hypothesis test_frequency_is_original(self=<pandas.tests.frame.test_apply.TestDataFrameAggregate instance at 0x7f2c3f1955a8>, index=DatetimeIndex(['2237-05-12', '2237-05-13', '2237-05-14', '2237-05-15'], dtype='datetime64[ns]', freq=None), num_columns=3) produces unreliable results: Falsified on the first call but did not on a subsequent one
../../../miniconda3/envs/pandas/lib/python2.7/site-packages/hypothesis/core.py:884: Flaky
```

File: pandas/tests/indexes/multi/test_monotonic.py [hunk 5]
```diff
@@ -182,22 +183,28 @@ def test_searchsorted_monotonic(indices):
     # test searchsorted only for increasing
     if indices.is_monotonic_increasing:
```
Comment (jreback) [inline pandas/tests/indexes/multi/test_monotonic.py]: pls don't import is_scalar from testing, import it from the canonical location

pandas.api.types.is_scalar.

File: pandas/util/testing.py [hunk 6]
```diff
@@ -28,6 +28,7 @@
     is_datetimelike_v_numeric, is_datetimelike_v_object,
     is_extension_array_dtype, is_interval_dtype, is_list_like, is_number,
     is_period_dtype, is_sequence, is_timedelta64_dtype, needs_i8_conversion)
+from pandas.core.dtypes.common import is_scalar  # noqa: F401
```
Comment (jreback) [inline pandas/util/testing.py]: don't add this

Comment (jreback) [reply pandas/util/testing.py]: these are internally used imports. this is not exporting these functions.

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
@topper-123 happy to have this if it can be done in a reasonbale way. pls update or close.

**Comment by topper-123** [comment]
I've updated the PR.

**Comment by topper-123** [comment]
Is thia ok? This would make the searchsorted method for all of the pandas objects return scalars for scalar inputs and make the API consistent.

**Comment by jreback** [comment]
this PR looked ok, merge master.

**Comment by topper-123** [comment]
Rebased.

File: pandas/core/base.py [hunk 7]
```diff
@@ -1350,8 +1350,14 @@ def factorize(self, sort=False, na_sentinel=-1):
 
         Returns
         -------
-        indices : array of ints
-            Array of insertion points with the same shape as `value`.
+        int or array of ints
+            A scalar or array of insertion points with the
+            same shape as `value`.
```
Comment (jreback) [inline pandas/core/base.py]: @datapythonista is this the correct format?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/base.py (hunk 7 above)
Comment (datapythonista) [reply pandas/core/base.py]: looks good, but can you use `int or array of int` (I'd like to parse at some point the types from these, so I prefer the type name `int` over `ints`)

File: pandas/core/base.py [hunk 8]
```diff
@@ -1338,8 +1338,8 @@ def factorize(self, sort=False, na_sentinel=-1):
 
         Parameters
         ----------
-        value : array_like
-            Values to insert into `self`.
+        value : scalar or array_like
+            Value(s) to insert into `self`.t
```
Comment (datapythonista) [inline pandas/core/base.py]:
```suggestion
             Value(s) to insert into ``self.t``.
```
I think render the whole `self.t` as code makes more sense (also, we end parameter descriptions with a period).

Comment (topper-123) [reply pandas/core/base.py]: I think this is actually better: ``Value(s) to search for in ``self``.``

As we're nor actually inserting, but searching.

**Review by jreback — APPROVED** [review]
